### ISSUE TITLE
Rotating with Expression Animation

### ISSUE THREAD

**Issue by Brian-Ding** [Microsoft/WindowsUIDevLabs#157]
Hey guys,

I would like a rectangle to keep rotating align with current second ( DateTime.Now.Second ), and here is my code. According to the page 7 [ in this slide](https://view.officeapps.live.com/op/view.aspx?src=https%3a%2f%2fsec.ch9.ms%2fsessions%2fbuild%2f2016%2fP486.pptx), I am assuming that GetTime() will be called repeated automatically, but it seems not that case. GetTime() will be called only once when I navigate to the MainPage.

 How can I achieve that? or maybe any other way around? Thx!

        Compositor _compositor;
        Visual _visual;
        CompositionPropertySet _sharedProperties;

        public MainPage()
        {
            this.InitializeComponent();

            _compositor = ElementCompositionPreview.GetElementVisual(this).Compositor;
            _visual = ElementCompositionPreview.GetElementVisual(ellipse);
            _visual.RotationAxis = new Vector3(0, 0, 1f);
            _visual.Size = new Vector2(50, 50);
            _visual.CenterPoint = new Vector3(25, 25, 0);

            _sharedProperties = _compositor.CreatePropertySet();
            _sharedProperties.InsertScalar("second", GetTime());

            PrepareAnimation();
        }

        private void PrepareAnimation()
        {
            var rotateAnimation = _compositor.CreateExpressionAnimation("props.second * 6");
            rotateAnimation.SetReferenceParameter("props", _sharedProperties);

            _visual.StartAnimation("RotationAngleInDegrees", rotateAnimation);
        }

        private float GetTime()
        {
            return (float)DateTime.Now.Second;
        }

**Comment by robmikh** [comment]
I'm afraid you're pulling the wrong conclusions from that sample. The sample is trying to highlight the ability to create custom properties in a `CompositionPropertySet` and update them from your app whenever you need to. The example method "GetHourOfDay" is only run the app calls it, the system compositor *will not call the method on its own*. The comment above that line attempts to clarify this, by stating that the "app will update this property when needed".  This means that you can update the value in the property set by calling `InsertScalar` whenever you need to, and the expression will update with the new value. There isn't a composition mechanism to have a property updated automatically using the return value of a method from your application.

Fortunately all you need to do to get your code to work is to update the property set on a timer. Because composition objects can be called from any thread, you can do this with the `System.Threading.Timer` object. You can find a GitHub issue [here](https://github.com/Microsoft/WindowsUIDevLabs/issues/34) where this approach was covered.

In practice it would look something like this:
```
private void StartLoop()
{
    _timer = new Timer(Loop, null, 1000, Timeout.Infinite);
}

private void Loop(object state)
{
    _sharedProperties.InsertScalar("second", GetTime());
    _timer.Change(1000, Timeout.Infinite);
}
```

**Comment by Brian-Ding** [comment]
Much Appreciated! Sorry for misunderstanding there, and thank you for clarifying that.
And I am have another issue, I would like an ellipse always moving toward the mouse/finger/pen. and I come up with the code below

        Compositor _compositor;
        Visual _visual;
        CompositionPropertySet _sharedProperties;
        Vector3KeyFrameAnimation offsetAnimation;

        public MainPage()
        {
            this.InitializeComponent();
            this.PointerMoved += MainPage_PointerMoved;

            _compositor = ElementCompositionPreview.GetElementVisual(this).Compositor;
            _visual = ElementCompositionPreview.GetElementVisual(ellipse);
           
            _sharedProperties = _compositor.CreatePropertySet();
            _sharedProperties.InsertVector3("delta", new Vector3(0, 0, 0));

            PrepareAnimation();
        }

        private void MainPage_PointerMoved(object sender, PointerRoutedEventArgs e)
        {
            Vector3 position = new Vector3((float)e.GetCurrentPoint(ellipse).Position.X, (float)e.GetCurrentPoint(ellipse).Position.Y, 0);
            _sharedProperties.InsertVector3("delta", position);

            xText.Text = $"Position X: {e.GetCurrentPoint(ellipse).Position.X}\tOffset X: {_visual.Offset.X}";
            yText.Text = $"Position X: {e.GetCurrentPoint(ellipse).Position.Y}\tOffset X: {_visual.Offset.Y}";

            _visual.StartAnimation("Offset", offsetAnimation);
        }

        private void PrepareAnimation()
        {
            offsetAnimation = _compositor.CreateVector3KeyFrameAnimation();
            offsetAnimation.InsertExpressionKeyFrame(0.0f, "this.StartingValue");
            offsetAnimation.InsertExpressionKeyFrame(1.0f, "this.StartingValue + props.delta");
            offsetAnimation.SetReferenceParameter("props", _sharedProperties);
            offsetAnimation.Duration = TimeSpan.FromSeconds(1);
        }

The problem is the ellipse will only move toward the mouse when I stop the mouse.  Is there any chance you might have some ideas about that?

**Comment by Brian-Ding** [comment]
Now I can make the rectangle follow cursor immediately with the code below,

        Compositor _compositor;
        Visual _visual;
        CompositionPropertySet _sharedProperties;
        ExpressionAnimation offsetAnimation;

        public MainPage()
        {
            this.InitializeComponent();
            this.PointerMoved += MainPage_PointerMoved;

            _compositor = ElementCompositionPreview.GetElementVisual(this).Compositor;
            _visual = ElementCompositionPreview.GetElementVisual(ellipse);
            _visual.Size = new Vector2(10, 10);
            _visual.AnchorPoint = new Vector2(0.5f, 0.5f);

            _sharedProperties = _compositor.CreatePropertySet();
            _sharedProperties.InsertVector3("delta", new Vector3(0, 0, 0));
            _sharedProperties.InsertVector3("currentPosition", _visual.Offset);

            PrepareAnimation();
        }

        private void MainPage_PointerMoved(object sender, PointerRoutedEventArgs e)
        {
            Vector3 delta = new Vector3((float)e.GetCurrentPoint(this).Position.X, (float)e.GetCurrentPoint(this).Position.Y, 0);
            _sharedProperties.InsertVector3("delta", delta);
        }

        private void PrepareAnimation()
        {
            offsetAnimation = _compositor.CreateExpressionAnimation();
            offsetAnimation.Expression = "props.delta";
            offsetAnimation.SetReferenceParameter("props", _sharedProperties);

            _visual.StartAnimation("Offset", offsetAnimation);
        }

But the rectangle is more like following immediately, no animation can be seen. I would like the rectangle to move toward cursor slowly.

**Comment by msftdavid** [comment]
@Brian-Ding Are you still having any of the issues you've mentioned?

**Comment by Brian-Ding** [comment]
@msftdavid I am able to make the rectangle always in position with cursor, but having problem to make the rectangle move toward cursor whenever I am moving the cursor or leave it still

**Comment by robmikh** [comment]
Did you want your rectangle to trail behind the cursor (taking the same path)? Or slowly move towards the cursor in a straight line?

**Comment by Brian-Ding** [comment]
Not sure if this is possible, but the ideal case is that: 
1. As I move the cursor, rectangle will be able to change its moving direction in time accordingly. ( Always moving towards cursor). In every short time, I think it is a path of straight line.
2. If I top the cursor later, rectangle will finally be in the same position with cursor.

**Comment by robmikh** [comment]
Hmmm... this seems to be a pretty special case. Someone else can correct me if I'm wrong, but I don't think you'll be able to do this with expressions or animations. You'll need to use a timer to update the visual's position based on the position of the cursor. I rigged up a really crude example:

```
private Compositor _compositor;
private SpriteVisual _visual;

private const int TIMERINTERVAL = 16;
private const float RATECHANGE = 15;
private Vector2 _currentCursorLocation;
private object _locationLock;
private Timer _timer;

public MainPage()
{
    this.InitializeComponent();

    InitComposition();
    InitCustomTracker();
    Window.Current.CoreWindow.PointerMoved += CoreWindow_PointerMoved;
}

private void InitComposition()
{
    _compositor = ElementCompositionPreview.GetElementVisual(this).Compositor;
    _visual = _compositor.CreateSpriteVisual();

    _visual.Size = new Vector2(75, 75);
    _visual.Brush = _compositor.CreateColorBrush(Colors.Red);

    ElementCompositionPreview.SetElementChildVisual(this, _visual);
}

private void InitCustomTracker()
{
    _locationLock = new object();
    _timer = new Timer(UpdateCursor, null, TIMERINTERVAL, Timeout.Infinite);
}

private void UpdateCursor(object state)
{
    var stopwatch = new Stopwatch();
    stopwatch.Start();

    bool needChangeX = false;
    bool needChangeY = false;
    float changeX = 0;
    float changeY = 0;

    // First let's get the cursor
    Vector2 cursorLocation;
    Vector2 cursorLocationRaw;
    lock(_locationLock)
    {
        cursorLocationRaw = _currentCursorLocation;
    }
    cursorLocation = new Vector2((int)cursorLocationRaw.X, (int)cursorLocationRaw.Y);
    var visualLocation = new Vector2((int)_visual.Offset.X, (int)_visual.Offset.Y);

    // Now let's calculate X
    if (cursorLocation.X > visualLocation.X)
    {
        if ((cursorLocation.X - visualLocation.X) < RATECHANGE)
        {
            _visual.Offset = new Vector3(cursorLocationRaw.X, _visual.Offset.Y, 0);
        }
        else
        {
            changeX = RATECHANGE;
            needChangeX = true;
        }
    }
    else if (cursorLocation.X < visualLocation.X)
    {
        if ((visualLocation.X - cursorLocation.X ) < RATECHANGE)
        {
            _visual.Offset = new Vector3(cursorLocationRaw.X, _visual.Offset.Y, 0);
        }
        else
        {
            changeX = -1.0f * RATECHANGE;
            needChangeX = true;
        }
    }
    // else do nothing

    // Now let's calculate Y
    if (cursorLocation.Y > visualLocation.Y)
    {
        if ((cursorLocation.Y - visualLocation.Y) < RATECHANGE)
        {
            _visual.Offset = new Vector3(_visual.Offset.X, cursorLocationRaw.Y, 0);
        }
        else
        {
            changeY = RATECHANGE;
            needChangeY = true;
        }
    }
    else if (cursorLocation.Y < visualLocation.Y)
    {
        if ((visualLocation.Y - cursorLocation.Y) < RATECHANGE)
        {
            _visual.Offset = new Vector3(_visual.Offset.X, cursorLocationRaw.Y, 0);
        }
        else
        {
            changeY = -1.0f * RATECHANGE;
            needChangeY = true;
        }
    }
    // else do nothing

    // Let's update our visual
    if (needChangeX || needChangeY)
    {
        var visualOffset = _visual.Offset;

        if (needChangeX)
        {
            visualOffset.X += changeX;
        }

        if (needChangeY)
        {
            visualOffset.Y += changeY;
        }

        _visual.Offset = visualOffset;
    }

    _timer.Change((int)Math.Max(0, TIMERINTERVAL - stopwatch.ElapsedMilliseconds), Timeout.Infinite);
}

private void CoreWindow_PointerMoved(CoreWindow sender, PointerEventArgs args)
{
    lock(_locationLock)
    {
        _currentCursorLocation = args.CurrentPoint.Position.ToVector2();
    }
}
```

This will follow the cursor, but it looks kind of weird since there are no attempts to interpolate between values or to switch direction gracefully. I'd highly recommend doing so if you wanted to build something like this. The code is also probably not as clean as it could be, but it does highlight a view key issues you'll need to tackle:

- If you use the System.Threading.Timer object, then you'll need a way to synchronize updates to the cursor position between the UI thread and the thread that the timer will call back on. I just used a simple lock here.
- You'll need to do some fuzzy comparisons between the visual and the cursor position. Because of floating point precision, you'll need to make sure you deal with values that are "close enough" in the real world. This is why the code has this weird integer rounding scheme. Without it the visual would bounce violently when the cursor was at certain locations. There are certainly better was to do this, but this was a quick and dirty proof of concept, so I reserved the right to be sloppy. :)
- Make sure you don't overshoot the cursor, or else you could introduce violent bouncing as well. This is why the code does comparisons to make sure the difference between the cursor and the visual is not under the set rate of change.

In the end, I'd actually recommend picking up some reading material around simple game AI. This is a common problem that many people have tackled (you might find it under trying to get an NPC to follow the player, for example). I'm sure they have a much more elegant solution than what I've cooked up. :)

**Comment by Brian-Ding** [comment]
What a detailed explanation and code above! And also thank you for pointing out the direction that I might work on. I will see what I can get from there.
